pkg/storage/utils/decomposedfs/decomposedfs.go

```diff
@@ -126,21 +131,108 @@ func NewDefault(m map[string]interface{}, bs tree.Blobstore) (storage.FS, error)
 // New returns an implementation of the storage.FS interface that talks to
 // a local filesystem.
 func New(o *options.Options, lu *lookup.Lookup, p PermissionsChecker, tp Tree, permissionsClient CS3PermissionsClient) (storage.FS, error) {
+	log := logger.New()
 	err := tp.Setup()
 	if err != nil {
-		logger.New().Error().Err(err).
-			Msg("could not setup tree")
+		log.Error().Err(err).Msg("could not setup tree")
 		return nil, errors.Wrap(err, "could not setup tree")
 	}
 
-	return &Decomposedfs{
+	var ev events.Stream
+	if o.Events.NatsAddress != "" {
+		ev, err = server.NewNatsStream(natsjs.Address(o.Events.NatsAddress), natsjs.ClusterID(o.Events.NatsClusterID))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	fs := &Decomposedfs{
 		tp:                tp,
 		lu:                lu,
 		o:                 o,
 		p:                 p,
 		chunkHandler:      chunking.NewChunkHandler(filepath.Join(o.Root, "uploads")),
 		permissionsClient: permissionsClient,
-	}, nil
+		stream:            ev,
+	}
+
+	if o.AsyncFileUploads {
+		if o.Events.NatsAddress == "" {
+			log.Error().Msg("need nats for async file processing")
+			return nil, errors.New("need nats for async file processing")
+		}
+
+		ch, err := events.Consume(ev, "dcfs", events.PostprocessingFinished{}, events.VirusscanFinished{})
+		if err != nil {
+			return nil, err
+		}
+
+		go fs.Postprocessing(ch)
+	}
+
+	return fs, nil
+}
+
+// Postprocessing starts the postprocessing result collector
+func (fs *Decomposedfs) Postprocessing(ch <-chan interface{}) {
+	ctx := context.TODO()
+	log := logger.New()
+	for event := range ch {
+		switch ev := event.(type) {
+		case events.PostprocessingFinished:
+			up, err := upload.Get(ctx, ev.UploadID, fs.lu, fs.tp, fs.o.Root, fs.stream, fs.o.AsyncFileUploads, fs.o.Tokens)
+			if err != nil {
+				log.Error().Err(err).Str("uploadID", ev.UploadID).Msg("Failed to get upload")
+				continue // NOTE: since we can't get the upload, we can't delete the blob
+			}
+
+			var (
+				failed     bool
+				keepUpload bool
+			)
+
+			switch ev.Action {
+			default:
+				log.Error().Str("action", ev.Action).Str("uploadID", ev.UploadID).Msg("unknown postprocessing outcome - aborting")
+				fallthrough
+			case "abort":
+				failed = true
+				keepUpload = true
+			case "continue":
+				if err := upload.Finalize(up); err != nil {
+					log.Error().Err(err).Str("uploadID", ev.UploadID).Msg("could not finalize upload")
+					keepUpload = true // should we keep the upload when assembling failed?
+					failed = true
+				}
+			case "delete":
+				failed = true
+			}
+
+			upload.Cleanup(up, failed, keepUpload)
+
+			if err := events.Publish(fs.stream, events.UploadReady{UploadID: ev.UploadID}); err != nil {
+				log.Error().Err(err).Str("uploadID", ev.UploadID).Msg("Failed to publish UploadReady event")
+			}
+		case events.VirusscanFinished:
```

Comment:
not too happy that the reva postprocessing knows about the virus service, fine for now but would be nice to have some more generic solutions.

Reply:
Yes that would be better. But I'm not sure how a more generic event should look like, so I opt keeping it until we have more events and can deduct a reasonable generic event

Reply:
@butonic 
Not entirely. Reva needs to update the node with the postprocessing result. E.g. last virusscan date. This information is important for the user. It could do this in two ways:
- Wait for the `PostprocessingFinished` event and update data for all steps at once
- Update the node when each postprocessing step is finished
The second option is imo superior as clients could potentially show the status of the postprocessing

That being said, as @fschade mentioned above already, it would be better to have a generic event like `PostprocessingStepFinished` on which reva could update the node. But as I explained above I would wait for that until we have more postprocessing steps and then design this event according to common values